File: benchmarks/R/harness.R
# try to follow style guide: http://r-pkgs.had.co.nz/style.html
source("benchmark.R")

Run <- function() {
  structure(list(
      name             ="",
      benchmark_suite  = NULL,
      num_iterations   = 1,
      inner_iterations = 1,
      total            = 0), 
    class = "Run")
}

run_benchmark <- function(r) {
  writeLines(paste0("Starting ", r$name, " benchmark ..."))
  r <- do_runs(r, r$benchmark_suite())
  report_benchmark(r)
  writeLines("")
}

measure <- function(r, bench) {
  # TODO: ask for platform independent way of measuring milliseconds/microseconds
  start_time <- proc.time()[[3]]
  if (!inner_benchmark_loop(bench, r$inner_iterations)) {
    stop(simpleError("Benchmark failed with incorrect result"))
  }
  end_time <- proc.time()[[3]]
  run_time <- (end_time - start_time) * 1000 * 1000 # we need to report microseconds
  print_result(r, run_time)
  r$total <- r$total + run_time
  r
}

do_runs <- function(r, bench) {
  for (i in 1:r$num_iterations) {
    r <- measure(r, bench)
  }
  r
}

report_benchmark <- function (r) {
  writeLines(paste0(r$name, ": iterations=", r$num_iterations,
                    " average: ", r$total / r$num_iterations, "us total: ",
                    r$total, "us\n"))
}

print_result <- function (r, run_time) {
  writeLines(paste0(r$name, ": iterations=1 runtime: ", runTime, "us"))
}

print_total <- function (r) {
  writeLines(paste0("Total Runtime: ", r$total, "us"))
}

get_suite_from_name <- function(benchmark_name) {
  filename <- paste0(tolower(benchmark_name), ".R")
  p <- source(filename)
  p$value
}

process_arguments <- function(args, run) {
  run$name <- args[1]
  run$benchmark_suite <- get_suite_from_name(args[1])
  
  if (length(args) > 1) {
    run$num_iterations <- strtoi(args[2])
    if (length(args) > 2) {
      run$inner_iterations <- strtoi(args[3])
    }
  }
  run
}

print_usage <- function() {
  writeLines("Harness [benchmark] [num-iterations [inner-iter]]")
  writeLines("")
  writeLines("  benchmark      - benchmark class name ")
  writeLines("  num-iterations - number of times to execute benchmark, default: 1")
  writeLines("  inner-iter     - number of times the benchmark is executed in an inner loop, ")
  writeLines("                   which is measured in total, default: 1")
}


args <- commandArgs(trailingOnly = TRUE)
args <- c("NBody")
if (length(args) < 1) {
  print_usage()
  quit(status = 1)
}

run <- Run()
run <- process_arguments(args, run)
run <- run_benchmark(run)
print_total(run)

File: benchmarks/CSharp/Harness.cs
using System.Diagnostics;

public class Run
{
  public readonly string Name;
  public readonly Type BenchmarkSuite;
  public int NumIterations;
  public int InnerIterations;
  public long Total;

  public Run(String name)
  {
    this.Name = name;
    this.BenchmarkSuite = SuiteFromName(name);
    NumIterations = 1;
    InnerIterations = 1;
  }

  public Type SuiteFromName(String name)
  {
    Type? suite = Type.GetType(name);
    if (suite == null) {
        throw new Exception("Suite " + name + " not found");
    }
    return suite;
  }


  public void RunBenchmark()
  {
    Console.WriteLine("Starting " + Name + " benchmark ...");

    Benchmark benchmark = (Benchmark) Activator.CreateInstance(BenchmarkSuite);

    DoRuns(benchmark);

    ReportBenchmark();

    Console.WriteLine("");
  }

  public void measure(Benchmark bench)
  {
    Stopwatch timer = Stopwatch.StartNew();
    if (!bench.InnerBenchmarkLoop(InnerIterations))
    {
      throw new InvalidOperationException("Benchmark failed with incorrect result");
    }
    timer.Stop();
    long runTime = timer.ElapsedMilliseconds;

    PrintResult(runTime);

    Total += runTime;
  }

  public void DoRuns(Benchmark bench)
  {
    for (int i = 0; i < NumIterations; i++)
    {
      measure(bench);
    }
  }

  public void ReportBenchmark()
  {
    Console.WriteLine(Name + ": iterations=" + NumIterations +
        " average: " + (Total / NumIterations) + "ms total: " + Total + "ms/n");
  }

  public void PrintResult(long runTime)
  {
    Console.WriteLine(Name + ": iterations=1 runtime: " + runTime + "ms");
  }

  public void PrintTotal()
  {
    Console.WriteLine("Total Runtime: " + Total + "ms");
  }
}

public class Harness
{
  public static Run ProcessArguments(String[] args)
  {
    Run run = new Run(args[0]);

    if (args.Length > 1)
    {
      run.NumIterations = Int32.Parse(args[1]);
      if (args.Length > 2)
      {
        run.InnerIterations = Int32.Parse(args[2]);
      }
    }
    return run;
  }

  public static void PrintUsage()
  {
    Console.WriteLine("Harness [benchmark] [num-iterations inner-iter]]");
    Console.WriteLine(" ");
    Console.WriteLine(" benchmark       -benchmark class name");
    Console.WriteLine(" num-iterations  -number of times to execute benchmark, default: 1");
    Console.WriteLine(" inner-iter      -number of times the benchmark is executed in an inner loop, ");
    Console.WriteLine("                  which is measured in total, default: 1");
  }

  public static int Main(String[] args)
  {
    if (args.Length < 2)
    {
      PrintUsage();
      return 1;
    }

    Run run = ProcessArguments(args);
    run.RunBenchmark();
    run.PrintTotal();
    return 0;
  }
}